lib/spack/spack/binary_distribution.py

```diff
@@ -2051,20 +2061,22 @@ def _tar_strip_component(tar: tarfile.TarFile, prefix: str):
     # Including trailing /, otherwise we end up with absolute paths.
     regex = re.compile(re.escape(prefix) + "/*")
 
-    # Remove the top-level directory from the member (link)names.
+    # Only yield members in the package prefix.
     # Note: when a tarfile is created, relative in-prefix symlinks are
     # expanded to matching member names of tarfile entries. So, we have
     # to ensure that those are updated too.
     # Absolute symlinks are copied verbatim -- relocation should take care of
     # them.
     for m in tar.getmembers():
         result = regex.match(m.name)
-        assert result is not None
+        if not result:
+            continue
         m.name = m.name[result.end() :]
         if m.linkname:
             result = regex.match(m.linkname)
             if result:
                 m.linkname = m.linkname[result.end() :]
+        yield m
```

Comment:
Can we update the docstring to reflect this is now a generator passed to `tar.extractall`?